host.go

```diff
@@ -695,4 +698,202 @@ func (h *Host) InitCommands(c *chat.Commands) {
+		}
+		return h.auth.LoadWhitelistFromFile(h.auth.whitelistFile)
+	}
+
+	allowlistReverify := func(room *chat.Room) []string {
+		if !h.auth.WhitelistMode() {
+			return []string{"allowlist is disabled, so nobody will be kicked"}
+		}
+		var kicked []string
+		forConnectedUsers(func(user *chat.Member, pk ssh.PublicKey) error {
+			if h.auth.CheckPublicKey(pk) != nil && !user.IsOp { // we do this check here as well for ops without keys
+				kicked = append(kicked, user.Name())
+				user.Close()
+			}
+			return nil
+		})
+		if kicked != nil {
+			room.Send(message.NewAnnounceMsg(fmt.Sprintf("%v were kicked during pubkey reverification.", kicked)))
+		}
+		return nil
+	}
+
+	allowlistStatus := func() (msgs []string) {
+		if h.auth.WhitelistMode() {
+			msgs = []string{"The allowlist is currently enabled."}
+		} else {
+			msgs = []string{"The allowlist is currently disabled."}
+		}
+		whitelistedUsers := []string{}
+		whitelistedKeys := []string{}
+		h.auth.whitelist.Each(func(key string, item set.Item) error {
+			keyFP := item.Key()
+			if forConnectedUsers(func(user *chat.Member, pk ssh.PublicKey) error {
+				if pk != nil && sshd.Fingerprint(pk) == keyFP {
+					whitelistedUsers = append(whitelistedUsers, user.Name())
+					return errors.New("not an actual error, but exit early because we found the key")
+				}
+				return nil
+			}) == nil {
+				// if we land here, the key matches no users
+				whitelistedKeys = append(whitelistedKeys, keyFP)
+			}
+			return nil
+		})
+		if len(whitelistedUsers) != 0 {
+			msgs = append(msgs, fmt.Sprintf("The following connected users are on the allowlist: %v", whitelistedUsers))
+		}
+		if len(whitelistedKeys) != 0 {
+			msgs = append(msgs, fmt.Sprintf("The following keys of not connected users are on the allowlist: %v", whitelistedKeys))
+		}
+		return
+	}
+
+	c.Add(chat.Command{
+		Op:         true,
+		Prefix:     "/allowlist",
+		PrefixHelp: "COMMAND [ARGS...]",
+		Help:       "Modify the allowlist or allowlist state. See /allowlist help for subcommands",
+		Handler: func(room *chat.Room, msg message.CommandMsg) (err error) {
+			if !room.IsOp(msg.From()) {
+				return errors.New("must be op")
+			}
+
+			args := msg.Args()
+			if len(args) == 0 {
+				args = []string{"help"}
+			}
+
+			// send exactly one message to preserve order
+			var replyLines []string
+
+			switch args[0] {
+			case "help":
+				replyLines = allowlistHelptext
+			case "on":
+				h.auth.SetWhitelistMode(true)
+			case "off":
+				h.auth.SetWhitelistMode(false)
+			case "add":
+				replyLines = forPubkeyUser(args[1:], func(pk ssh.PublicKey) { h.auth.Whitelist(pk, 0) })
+			case "remove":
+				replyLines = forPubkeyUser(args[1:], func(pk ssh.PublicKey) { h.auth.Whitelist(pk, 1) })
+			case "import":
+				replyLines, err = allowlistImport(args[1:])
+			case "reload":
+				err = allowlistReload(args[1:])
+			case "reverify":
+				replyLines = allowlistReverify(room)
+			case "status":
+				replyLines = allowlistStatus()
```

Comment:
All the calls made by these methods on h.auth struct should be protected by the mutex, right?

eg. `allowlistImport` calls `h.auth.Whitelist` which updates the `whitelist` set in the auth object, shouldn't that be done with the mutex locked?

Reply:
As far as I understand, the set has its own internal mutex.

Reply:
Unless you mean that importing etc. should be one atomic action.